src/topology/homotopy/H_spaces.lean

```diff
@@ -0,0 +1,498 @@
+begin
+  apply continuous.piecewise,
+  { rintros ⟨_, t⟩ h,
+    have h_eq : (λ (i : (path x x × path x x) × I), (i.snd : ℝ) ≤ (1 / 2)) =
+      (set.univ) ×ˢ {s : I | (s : ℝ) ≤ (1 / 2)},
+    { ext p,
+      change (p.2 : ℝ) ≤ 1 / 2 ↔ p ∈ (@set.univ (Ω(x) × Ω(x)) ×ˢ {s : I | (s : ℝ) ≤ 1 / 2}),
+      simp only [set.mem_prod, set.mem_univ, set.mem_set_of_eq, true_and] },
+    erw h_eq at h,
+    simp only [frontier_prod_eq, frontier_univ, closure_univ, set.empty_prod,
+      set.union_empty, set.prod_mk_mem_set_prod_eq, set.mem_univ, true_and] at h,
+    simp only [mem_frontier h, set.right_mem_Icc, zero_le_one, coe_one, one_div,
+      mul_inv_cancel_of_invertible, path.extend_extends, mk_one, path.target, set.left_mem_Icc,
+      sub_self, mk_zero, path.source] },
+  exacts [continuous_prod_first_half x, continuous_prod_second_half x],
+end
+
+/- This is the function defined on p. 475 of Serre's `Homologie singulière des espaces fibrés`
+defining a homotopy from a path `γ` to the product `γ ∧ e`.-/
+def delayed_refl_left {x : X} (θ : I) (γ : Ω(x)) : Ω(x) :=
+{ to_fun := λ t, if  (t : ℝ) ≤ θ / 2 then x
+                 else γ.extend ((2*t - θ)/(2 - θ)),
+  continuous_to_fun :=
+  begin
+    apply continuous.piecewise,
+    { intros t ht,
+      rw mem_frontier ht,
+      field_simp, },
+    exacts [continuous_const, (continuous_extend γ).comp ((continuous_const.mul
+      continuous_induced_dom).sub continuous_const).div_const ],
+  end,
+  source' :=
+  begin
+    simp only [coe_zero, path.source, mul_zero, zero_sub, ite_eq_left_iff, not_le],
+    intro h,
+    contrapose! h,
+    exact div_nonneg θ.2.1 zero_le_two,
+  end,
+  target' :=
+  begin
+    simp only [coe_one, path.target, mul_one, ite_eq_left_iff, not_le],
+    intro,
+    rw div_self,
+    { simpa only [div_self, set.right_mem_Icc, zero_le_one, extend_extends, mk_one,
+      to_fun_eq_coe, coe_to_continuous_map] using γ.3 },
+    { linarith },
+  end }
+
+
+lemma delayed_refl_left_at_0 (γ : Ω(x)) : (delayed_refl_left 0 γ) = γ :=
+begin
+  ext t,
+  dsimp [delayed_refl_left],
+  rw [sub_zero, sub_zero, mul_div_assoc, mul_div_cancel' (t : ℝ) two_ne_zero, zero_div],
+  split_ifs with h,
+  { have := coe_eq_zero.mp (((nonneg t).antisymm) h).symm,
+    rwa [this, path.source] },
+  { exact extend_extends' _ _ },
+end
+
+lemma delayed_refl_left_at_1 (γ : Ω(x)) : (delayed_refl_left 1 γ) = (path.refl x).trans γ :=
+begin
+  ext t,
+  dsimp [delayed_refl_left],
+  norm_num,
+  refl,
+end
+
+lemma delayed_refl_left_e (θ : I) : delayed_refl_left θ (refl x) = refl x :=
+begin
+  ext t,
+  dsimp [path.refl, delayed_refl_left],
+  split_ifs;
+  refl,
+end
+
+/- This is the function analogous to the one on p. 475 of Serre's `Homologie singulière des espaces`
+  `fibrés`, defining a homotopy from the product path `e ∧ γ` to `γ`.-/
+def delayed_refl_right {x : X} (θ : I) (γ : Ω(x)) : Ω(x) :=
+{ to_fun := λ t, if  (t : ℝ) ≤ (1 + θ) / 2 then γ.extend (2 * t/(1 + θ))
+                 else x,
```

Comment:
Why not define this using `delayed_refl_left` so you don't need to duplicate the work? Actually, it feels `delayed_refl_right` is easier to work with so maybe we should define `delayed_refl_left` in terms of `delayed_refl_right`; the relevant identity is `1 - 2(1-t)/(1+(1-θ)) = (2t-θ)/(2-θ)`, and you just need to use [path.symm](https://leanprover-community.github.io/mathlib_docs/topology/path_connected.html#path.symm) twice.